sys/initrd.c

```diff
@@ -0,0 +1,246 @@
+#include <errno.h>
+#include <common.h>
+#include <malloc.h>
+#include <vm.h>
+#include <vm_map.h>
+#include <stdc.h>
+#include <cpio.h>
+#include <initrd.h>
+#include <vnode.h>
+#include <mount.h>
+#include <mount.h>
+#include <linker_set.h>
+
+/* ramdisk related data that will be stored in v_data field of vnode */
+typedef struct cpio_node {
+  TAILQ_ENTRY(cpio_node) c_list;
+
+  dev_t c_dev; 
+  ino_t c_ino;
+  mode_t c_mode;
+  nlink_t c_nlink;
+  uid_t c_uid;
+  gid_t c_gid;
+  dev_t c_rdev;
+  off_t c_size;
+  time_t c_mtime;
+
+  const char *c_path;
+  void *c_data;
+  
+  vnode_t *c_vnode;
```

Comment:
An object pointed by `c_vnode` may be freed without notice. Because of that, inside `cpio_node` there must be no references to objects other than another `cpio_node`.